### PR TITLE
*: cache store address.

### PR DESCRIPTION
**Pull request by siddontang**
@ngaut @BusyJay

### CODE REVIEW

File: src/server/resolve.rs [hunk 1]
```diff
@@ -44,18 +45,32 @@ impl Display for Task {
 pub struct Runner<T: PdClient> {
     cluster_id: u64,
     pd_client: Arc<RwLock<T>>,
+    addrs: HashMap<u64, String>,
```
Comment (ngaut) [inline src/server/resolve.rs]: s/addrs/store_addrs/

**Comment by ngaut** [comment]
Rest LGTM.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/server/resolve.rs [hunk 2]
```diff
@@ -44,18 +45,32 @@ impl Display for Task {
...
-        let addr = store.get_address();
-        let sock = try!(util::to_socket_addr(addr));
+        // If we use docker and use host for store address, the real IP
+        // may be changed after service restarts, so here we just cache
+        // pd result and use to_socket_addr to get real socket address.
+        let sock = try!(util::to_socket_addr(&*addr));
         Ok(sock)
     }
+
+    fn get_address(&mut self, store_id: u64) -> Result<String> {
```
Comment (BusyJay) [inline src/server/resolve.rs]: I think we can return `&str` here.

Comment (siddontang) [inline src/server/resolve.rs]: can't compile.

**Comment by BusyJay** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL

File: tests/raftstore/server.rs [hunk 3]
```diff
@@ -108,6 +108,12 @@ impl Simulator for ServerCluster {
 
         let mut cfg = cfg;
 
+        // Now we cache the store address, so here we should re-use last
+        // listening address for the same store.
+        if let Some(addr) = self.addrs.get(&node_id) {
```
Comment (tiancaiamao) [inline tests/raftstore/server.rs]: is that the the purpose of this PR ?

Comment (siddontang) [inline tests/raftstore/server.rs]: must re-use.

Comment (tiancaiamao) [inline tests/raftstore/server.rs]: is there any drawback if we use the cached address, say, inconsistent with pd?

Comment (siddontang) [inline tests/raftstore/server.rs]: no need to get store address from pd every time, and in production, we can not change store address, so it is safe to cache it.

Comment (tiancaiamao) [inline tests/raftstore/server.rs]: LGTM

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**